### PR TITLE
SAGA: Support Korean IHNM

### PR DESCRIPTION
**Pull request by phcoder**

### CODE REVIEW

File: engines/saga/font.cpp [hunk 1]
```diff
@@ -653,17 +700,69 @@ void DefaultFont::outFont(const FontStyle &drawFont, const char *text, size_t co
...
+			uint16 full = ((leading & 0x7f) << 8) | trailing;
+			int initial = (full >> 10) & 0x1f;
+			int mid = (full >> 5) & 0x1f;
+			int fin = full & 0x1f;
+			int initidx = initial - 1;
+			static const int mididxlut[0x20] = {
+				-1, -1, 0, 1, 2, 3, 4, 5, -1, -1, 6, 7, 8, 9,
+				10, 11, -1, -1, 12, 13, 14, 15, 16, 17, -1, -1,
+				18, 19, 20, 21, -1, -1};
+			// TODO: check this.
```
Comment (bluegr) [inline engines/saga/font.cpp:725]: What needs to be checked here?

File: engines/saga/font.cpp [hunk 2]
```diff
@@ -653,17 +700,69 @@ void DefaultFont::outFont(const FontStyle &drawFont, const char *text, size_t co
...
+			static const int mid2inivariant[32] = {
+				0, 0, 1, 0, 0, 0, 0, 0,
+				0, 0, 0, 0, 0, 1, 3, 3,
+				0, 0, 3, 1, 1, 3, 3, 3,
+				0, 0, 1, 1, 3, 0, 0, 0,
+			};
+			int mididx = mididxlut[mid];
+			int finidx = fin >= 0x12 ? fin - 2 : fin - 1;
+			if (initial >= 0x15 || initidx < 0 || mididx < 0 || fin == 0 || fin == 0x12 || fin >= 0x1e) {
+				// TODO: non-jamo
```
Comment (bluegr) [inline engines/saga/font.cpp:735]: Jamo fonts are used in Korean. What's this for? And what needs to be done here?

File: engines/saga/font.cpp [hunk 3]
```diff
@@ -653,17 +700,69 @@ void DefaultFont::outFont(const FontStyle &drawFont, const char *text, size_t co
...
+
+			int midvariant = 0;
+
+			if (fin != 1)
+				midvariant += 2;
+			if (!(initial == 2 || initial == 3 || initial == 17))
+				midvariant++;
+
+			// TODO: variants slightly move final consonant.
+			// Idk enough to choose the right one
```
Comment (bluegr) [inline engines/saga/font.cpp:748]: I don't understand this. Can you please elaborate?

File: engines/saga/font.cpp [hunk 4]
```diff
@@ -672,105 +771,68 @@ void DefaultFont::outFont(const FontStyle &drawFont, const char *text, size_t co
...
-			charHeight = drawFont.header.charHeight;
-			rowLength = drawFont.header.rowLength;
-			bitmap = &drawFont.font[drawFont.fontCharEntry[c_code].index];
-			charTracking = drawFont.fontCharEntry[c_code].tracking;
+			c_code = FONT_CH_QMARK;
+#else
+			// Character code is not defined, but advance tracking
+			// ( Not defined if offset is 0, except for 33 ('!') which
+			//   is defined )
+			textPoint.x += drawFont.fontCharEntry[c_code].tracking;
```
Comment (bluegr) [inline engines/saga/font.cpp:823]: So this will be shown as a blank character, right? Is this what we want, or do we want to output ? in this case?

**Comment by sev-** [comment]
The current PR is a WIP. It is guesswork, as indicated in those comments. Please, do not merge until those TODOs are confirmed with the original.

File: engines/saga/font.cpp (hunk 2 above)
Comment (phcoder) [reply engines/saga/font.cpp]: Korean has also some non-Hangul characters like Hanja, pictograms and pseudographics. Compare it to English using digits, emojis and pseudographics

Comment (phcoder) [reply engines/saga/font.cpp]: Clarified

File: engines/saga/font.cpp (hunk 3 above)
Comment (phcoder) [reply engines/saga/font.cpp]: Removed this comment as this is now done

File: engines/saga/font.cpp (hunk 4 above)
Comment (phcoder) [reply engines/saga/font.cpp:839]: Yes, as an empty char. It's easier to just skip, But putting '?' is also an option

File: engines/saga/font.cpp (hunk 1 above)
Comment (phcoder) [reply engines/saga/font.cpp]: Nothing anymore

**Comment by phcoder** [comment]
> The current PR is a WIP. It is guesswork, as indicated in those comments. Please, do not merge until those TODOs are confirmed with the original.

Fixed and confirmed with the original

**Comment by sev-** [comment]
Excellent, thank you!

**Pull request merged**